Fix recurring schedules, e.g. Cron and Interval schedules

thenbrent opened the pull request:
This is a monster, late stage PR for 3.0.0. However, because there are breaking changes, and it resolves a number of discrepancies and gotchas in the recurring and cron actions & schedules, I think it's worth including in 3.0.0 instead of waiting for a 4.0.0, or shipping in a 3.x.x version.

---

Fixes #330 and associated issues - #256, #258, #291, #293 and #296.

---

As outlined in #330, there are a number of issues with recurring action that use either a Cron or Interval schedule.

This PR fixes recurring schedules, and in the process, refactors all schedules to avoid duplicate code and ambiguity of properties and parameters.

This patch introduces two important new base classes:

 * `ActionScheduler_Abstract_Schedule` and
 * `ActionScheduler_Abstract_RecurringSchedule`

These are then used to implement all other schedule types - cron, recurring/interval and simple. This helps unify the data on each schedule. For example, schedules used to have different property names referring to equivalent data like `ActionScheduler_IntervalSchedule::start_timestamp` and `ActionScheduler_SimpleSchedule::timestamp`.

Both of these timestamps actually referring to the date the action was scheduled, but due to ambiguity in naming, like "_start_" or "_date_" or "_first_", the data was sometimes used to refer to the scheduled date, and sometimes the first instance of a recurring schedule. While unifying the properties, their nomenclature has also been changed to avoid ambiguity.

While working on this, I also took the opportunity to:

* harden safeguards when next instances are scheduled to make only recurring action are ever rescheduled
* introduce a new `ActionScheduler_CanceledSchedule` class which takes advantage of the separation of scheduled / next dates on a schedule so that a cancelled action's scheduled date can be derived (and displayed in the admin)
* improve docblocks to better record any gotchas, like time slippage when rescheduling an instance of a recurring action
* add more unit tests, particularly to test cases of 2...n instances of a recurring action
* log issues rescheduling a recurring action

thenbrent commented:
I've rebased this on `version_3_0_0` now that #259 has been merged. One of the new benefits of this changeset I found while merging in that code is that it's possible to migrate _cancelled_ actions now, because we can access their scheduled date via the new `ActionScheduler_CanceledSchedule`. Previously, they were simply deleted from the old data store instead of being migrated.

rrennick commented:
@thenbrent This is looking good. Just a couple things:

Related to #331, we should start ensuring any new code merged has consistent spacing. The WP coding standards for spacing provide for good readability. I used the following to find some of the inconsistent spacing (in added code only)

```
$ git show dfb1f11a10fec59c9e6d590d7a2c037ff5eb6a60 | grep ^+ | grep \( | grep -v '( '
+	public function test_next_instance_of_interval_action() {
+		$random    = md5(rand());
+		$date      = as_get_datetime_object('12 hours ago');
+		$store     = ActionScheduler::store();
+		$claim = $store->stake_claim(10, $date);
+		$actions = $claim->get_actions();
+		$this->assertCount(1, $actions);
+		$fetched_action_id = reset($actions);
+		$fetched_action    = $store->fetch_action($fetched_action_id);
+		$store->release_claim($claim);
+		$runner->run();
+		$date = as_get_datetime_object('+1 day');
+		$claim = $store->stake_claim(10, $date);
+		$fetched_action_id = reset($actions);
+		$fetched_action    = $store->fetch_action($fetched_action_id);
+		$store->release_claim($claim);
+		$date = as_get_datetime_object('+1 day');
+		$claim = $store->stake_claim(10, $date);
+		$actions = $claim->get_actions();
+		$this->assertCount(1, $actions);
+		$fetched_action_id = reset($actions);
+		$fetched_action    = $store->fetch_action($fetched_action_id);
```

I'm getting a warning in WP CLI and in the debug log:

```
$ wp option get siteurl
Notice: ActionScheduler_Schedule_Deprecated::next is <strong>deprecated</strong> since version 3.0.0! Use ActionScheduler_Schedule_Deprecated::get_date() instead. in /wp-includes/functions.php on line 4435
https://ron.localhost
```

thenbrent commented:
> we should start ensuring any new code merged has consistent spacing

@rrennick good call. I've fixed up spacing and rebased in the fixes.

I ran `git show HEAD...a4934fcea90 | grep ^+ | grep \( | grep -v '( '` to review the whole diff.

> I'm getting a warning in WP CLI and in the debug log:

I can't see any remaining calls to `ActionScheduler_Schedule_Deprecated::next()`. Is that notice coming from [Action Scheduler Admin](https://github.com/Prospress/action-scheduler-admin) plugin?

rrennick commented:
> I can't see any remaining calls to ActionScheduler_Schedule_Deprecated::next(). Is that notice coming from Action Scheduler Admin plugin?

Yes. Other than a rebase this is good to go.

thenbrent commented:
Rebased, thanks @rrennick!
